Add `KeyPathIterable` protocol and synthesis.

dan-zheng opened the pull request:
The `KeyPathIterable` protocol represents types whose values provide
custom key paths to properties or elements.

For types that conform to `KeyPathIterable`, the compiler can synthesize
a default implementation of `allKeyPaths` based on the type's stored
properties.

---

`KeyPathIterable` enables generic machine learning optimizers and will
be used to replace `ParameterGroup`.

The `ParameterGroup` protocol requirements are not sufficiently general
for many use cases: it is limited to a single parameter type and does not
support joint iteration over parameters for multiple `ParameterGroup`
instances.

`KeyPathIterable` solves these problems. In addition, it is a generally
useful language feature that can model both static stored properties
and custom dynamic properties.

---

`KeyPathIterable.swift` currently lives in `stdlib/public/core` to enable
future usage in AutoDiff tests without dependency on TensorFlow module.

---

`KeyPathIterable` is designed by @rxwei and myself.

`KeyPathIterable` is inspired by:
* Various discussions about Swift key paths, in particular [this tweet](https://twitter.com/jckarter/status/868195828955062272)
* [Objective-C Key-Value Coding](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/KeyValueCoding/index.html), [lenses in functional programming languages](https://en.wikibooks.org/wiki/Haskell/Lenses_and_functional_references)

[Here](https://gist.github.com/rxwei/e316f7114b723ad69a30a3aba224ccb3) is an extended design for `KeyPathIterable` that separates functionality into two protocols: `StoredPropertyIterable` and `CustomKeyPathIterable`.
* `StoredPropertyIterable` models the static layout of structs and provides a static collection of key paths to all stored properties.
* `CustomKeyPathIterable` can model both static and dynamic structures, including collection types like `Array` and `Dictionary`.

dan-zheng committed:
Add `KeyPathIterable` protocol and synthesis.

The `KeyPathIterable` protocol represents types whose values provide
custom key paths to properties or elements.

For types that conform to `KeyPathIterable`, the compiler can synthesize
a default implementation of `allKeyPaths` based on the type's stored
properties.

---

`KeyPathIterable` enables generic machine learning optimizers and will
be used to replace `ParameterGroup`.

The `ParameterGroup` protocol requirements are not sufficiently general
for many use cases: it is limited to a single parameter type and does not
support joint iteration over parameters for multiple `ParameterGroup`
instances.

`KeyPathIterable` solves these problems. In addition, it is a generally
useful language feature that can model both static stored properties
and custom dynamic properties.

dan-zheng commented on stdlib/public/core/KeyPathIterable.swift:
```diff
@@ -0,0 +1,115 @@
+public extension KeyPathIterable {
+  /// An array of all custom key paths of this value and any custom key paths
+  /// nested within each of what this value's key paths refers to.
+  var recursivelyAllKeyPaths: [PartialKeyPath<Self>] {
```
The naming of `recursivelyAllKeyPaths` and the doc comment have room for improvement.

ZkHaider replied:
`nestedKeyPaths`?

dan-zheng replied:
Thanks for the suggestion! I feel `nestedKeyPaths` is not an ideal name for this property because it also contains top-level `allKeyPaths`. `nestedKeyPaths` sounds like it *doesn't* contain top-level properties (or at least leaves it ambiguous).

---

We considered `nestedKeyPaths` as not just a different name, but a different design option (where `allNestedKeyPaths` refers to `recursivelyAllKeyPaths` excluding top-level `allKeyPaths`, and `recursivelyAllKeyPaths` would be defined as `allKeyPaths + allNestedKeyPaths`).

We decided not to implement that option though, since we couldn't think of many use cases for "just recursively nested non-top-level key paths".

rxwei commented on lib/Sema/DerivedConformanceKeyPathIterable.cpp:
```diff
@@ -0,0 +1,136 @@
+
+bool DerivedConformance::canDeriveKeyPathIterable(NominalTypeDecl *nominal) {
+  // There must be at least one stored property.
+  // TODO: Improve. Require struct?
```
Yes, require struct for now. But it could also make sense for classes: each subclass implementation simply appends things to `super.allKeyPaths`.

dan-zheng replied:
Changed the requirement to "must be struct" in https://github.com/apple/swift/pull/21557/commits/9b3769281937327bae90be3a24937063a0c39a47.
Also added comment about supporting classes in the future.

dan-zheng commented on lib/Sema/DerivedConformanceKeyPathIterable.cpp:
```diff
@@ -0,0 +1,136 @@
+
+bool DerivedConformance::canDeriveKeyPathIterable(NominalTypeDecl *nominal) {
+  // There must be at least one stored property.
+  // TODO: Improve. Require struct?
```
remove comment

rxwei commented on stdlib/public/core/KeyPathIterable.swift:
```diff
@@ -0,0 +1,115 @@
+public extension KeyPathIterable {
+  /// Returns an array of all custom key paths of this value, to the specified
+  /// type.
+  @inlinable
```
There's no reason to serialize it right now. Let's remove `@inlinable` (and other occurrences) and reevaluate this later. When the synthesis can directly emit a flattened array literal, we should then serialize it.

rxwei replied:
That said, the required computed property (which you synthesize) should be `@inlinable`.

dan-zheng replied:
Added `@inlinable` to synthesized `allKeyPaths` in https://github.com/apple/swift/pull/21557/commits/a0635fa7f8314c7af117066780877c01d018dbfb. Verified that `@inlinable` exists in SILGen.

rxwei commented on stdlib/public/core/KeyPathIterable.swift:
```diff
@@ -0,0 +1,115 @@
+  }
+}
+
+extension KeyPathIterable {
```
public extension

dan-zheng replied:
Done in https://github.com/apple/swift/pull/21557/commits/9b3769281937327bae90be3a24937063a0c39a47

rxwei commented on lib/Sema/DerivedConformanceKeyPathIterable.cpp:
```diff
@@ -0,0 +1,134 @@
+        UnresolvedDotExpr(nominalTypeExpr, SourceLoc(), member->getFullName(),
+                          DeclNameLoc(), /*Implicit*/ true);
+    auto *keyPathExpr =
+        new (C) KeyPathExpr(SourceLoc(), dotExpr, nullptr, /*Implicit*/ true);
```
Woo! This is so simple.

rxwei reviewed (approved):
Great!

dan-zheng committed:
Address feedback from @rxwei.
- Remove `@inlinable` from `KeyPathIterable` extension methods.
- Enable `KeyPathIterable` synthesis for empty structs, add test.
- Add comment by @rxwei about supporting synthesis for classes in the future.

rxwei commented on lib/Sema/DerivedConformanceKeyPathIterable.cpp:
```diff
@@ -0,0 +1,134 @@
+using namespace swift;
+
+bool DerivedConformance::canDeriveKeyPathIterable(NominalTypeDecl *nominal) {
+  return !nominal->getStoredProperties().empty();
```
Why can't empty structs derive this conformance? `allKeyPaths` is just `[]` in that case and we should definitely derive it.

dan-zheng replied:
Good point! Added support for empty structs (and a test) in https://github.com/apple/swift/pull/21557/commits/9b3769281937327bae90be3a24937063a0c39a47

dan-zheng committed:
Address feedback from @rxwei.

Mark synthesized `allKeyPaths` declaration with `@inlinable`.
Verified that `@inlinable` is correctly added in SILGen:

```
struct Parameters : KeyPathIterable {
  @sil_stored var w: Float { get set }
  @sil_stored var b: Float { get set }
  init(w: Float, b: Float)
  typealias AllKeyPaths = [PartialKeyPath<Parameters>]
  @inlinable var allKeyPaths: [PartialKeyPath<Parameters>] { get }
}
```

rxwei commented:
@swift-ci please test tensorflow

dan-zheng commented:
Note: the synthesis condition for `KeyPathIterable` could be tightened to: "synthesize requirements only if no requirements have been user customized".

This would prevent errors like the following:
```swift
// error.swift
struct Test : KeyPathIterable {
  var x: Float
  // Question:
  //   What if I customize `AllKeyPaths` to be sth that conforms to
  //   `Collection` but not `ExpressibleByArrayLiteral`?
  //   And I let the compiler synthesize `allKeyPaths`?
  // Answer:
  //    The compiler doesn't crash, but produces an error because
  //    the synthesized body of `allKeyPaths` assumes that `AllKeyPaths`
  //    conforms to `ExpressibleByArrayLiteral`.
  typealias AllKeyPaths = UnsafeBufferPointer<PartialKeyPath<Test>>
}
```
```
error.swift:1:8: error: type 'Test' does not conform to protocol 'KeyPathIterable'
struct Test : KeyPathIterable {
       ^
error.swift:1:8: error: protocol 'KeyPathIterable' is broken; cannot derive conformance for type 'Test'
struct Test : KeyPathIterable {
       ^
```

Errors like this aren't unique to `KeyPathIterable` though, [`CaseIterable` has a similar issue.](https://bugs.swift.org/browse/SR-9582)
Since the compiler doesn't crash, I think this error can be handled in a follow-up PR.

rxwei commented:
Great. Thanks for the detailed note!

natecook1000 commented on stdlib/public/core/KeyPathIterable.swift:
```diff
@@ -0,0 +1,112 @@
+
+public extension KeyPathIterable {
+  var _allKeyPathsTypeErased: [AnyKeyPath] {
+    return allKeyPaths.map { $0 as AnyKeyPath }
```
Since key paths are classes, you can cast the array for these two instead of mapping:

```swift
return allKeyPaths as [AnyKeyPath]
```

(This is really cool functionality!)

rxwei replied:
It's not really directly convertible. I'm getting this error:

```console
/Users/rxwei/Development/Swift/swift-tf-source/swift/stdlib/public/core/KeyPathIterable.swift:57:24: error: 'Self.AllKeyPaths' is not convertible to '[AnyKeyPath]'; did you mean to use 'as!' to force downcast?
    return allKeyPaths as [AnyKeyPath]
           ~~~~~~~~~~~~^~~~~~~~~~~~~~~
                       as!
```

rxwei replied:
If that works on master, then it's probably because the changes haven't been merged. The tensorflow branch is a pretty old (2018-11-24).

natecook1000 replied:
Ah, you’re right — I’d missed that `allKeyPaths` isn’t an array here, it’s an instance of the associated type. Carry on!

rxwei replied:
Ah right. I forgot that too :)

natecook1000 commented on stdlib/public/core/KeyPathIterable.swift:
```diff
@@ -0,0 +1,112 @@
+extension Dictionary : KeyPathIterable {
+  public typealias AllKeyPaths = [PartialKeyPath<Dictionary>]
+  public var allKeyPaths: [PartialKeyPath<Dictionary>] {
+    return keys.map { \Dictionary[$0]! }
```
Seems like this (and really all collections) should follow the array example of giving access via `indices`, yielding key paths to key-value pairs here. If you make `Dictionary.Values` key-path iterable, too, a user could write `myDictionary.values.allKeyPaths` to get this result.

rxwei replied:
Good point. Key paths using `subscript` and `Dictionary.Index` currently causes a crasher on our branch. This may have been fixed on master. We'll make sure to update these once we merge.
